jetty-ee10/jetty-ee10-annotations/src/main/java/org/eclipse/jetty/ee10/annotations/AnnotationParser.java

```diff
@@ -520,6 +531,44 @@ public FieldVisitor visitField(final int access,
         }
     }
 
+    /**
+     * Check whether the classname is located inside its containingResource at
+     * the location that matches its package. The comparison accounts for classes
+     * located within WEB-INF/classes.
+     * @param containingResource the Resource that contains the class
+     * @param classResource the Resource representing the class
+     * @param classname the package-qualified name of the class
+     * @return true if the containingResource contains a class file at the location matching the package name of the class
+     */
+    public static boolean checkClassContainment(Resource containingResource, Resource classResource, String classname)
+    {
+        if (containingResource == null || classResource == null)
+            return false;
+        Path relative = containingResource.getPathTo(classResource);
+        if (relative == null)
+            return false; //unable to be verified
+
+        StringTokenizer tokenizer = new StringTokenizer(classname, "/", false);
+
+        for (int i = 0; i < relative.getNameCount(); i++)
+        {
+            String s = relative.getName(i).toString();
+            if (i == 0 && "WEB-INF".equalsIgnoreCase(s))
+                continue;
+            if (i == 1 && "classes".equalsIgnoreCase(s))
+                continue;
+            if (i == relative.getNameCount() - 1)
+            {
+                if ("class".equals(FileID.getExtension(s)))
+                    s = s.substring(0, s.length() - 6);
+            }
```

Comment:
A comment saying what this is doing would be good.   something like `// strip the .class suffix from the last path item`.
Also, if the `.class` is missing then we should not match
Finally, it should be an equalsIgnoreCase

Reply:
This could be changed to ...

``` java
// strip class extensions
if (FileID.isExtension(s, "class"))
    s = s.substring(0, s.length() - 6)
```

As that will verify the extension in a case insensitive way.